### PR TITLE
Fix on-chain SPV non-determinism bug

### LINKED ISSUE #9: Non-deterministic floating point arithmetic in on-chain SPV
The SPV client currently stores results of floating point arithmetic in the store, causing non-determinism. The testnet produced 2 different encodings on x86-64 devices vs an ARM64 device.

Example (note the `log2work` field):

x86-64:
```json
'{"stored":{"header":{"version":549453824,"prev_blockhash":"0000000000005bc09e12caaa9e46d7af870d02e4ae691c210410bdaceaf566d9","merkle_root":"922a79b838e084e5663b8be5330cdc0ded713e1ed2b349416ec219e81d8eaf5e","time":1576210713,"bits":436449004,"nonce":1841051496},"height":1612364,"log2work":64.37285314603291},"id":"000000000000004267d6e454c1597e9db80b64ab2123ef0441de7892f837faed"}'
```

arm64:
```json
{"stored":{"header":{"version":549453824,"prev_blockhash":"0000000000005bc09e12caaa9e46d7af870d02e4ae691c210410bdaceaf566d9","merkle_root":"922a79b838e084e5663b8be5330cdc0ded713e1ed2b349416ec219e81d8eaf5e","time":1576210713,"bits":436449004,"nonce":1841051496},"height":1612364,"log2work":64.37285314603292},"id":"000000000000004267d6e454c1597e9db80b64ab2123ef0441de7892f837faed"}
```

This was easily tracked down because the floating point value was directly serialized in the store, but we should learn from this by never doing any floating point operations at all in on-chain logic.

We can fix this immediate issue by either replacing that `log2work` field and its operations with a u64, or by implementing a simpler SPV module which is designed with determinism in mind.

### PR DESCRIPTION
**Pull request by keppel**
Instead of storing log2(work) as an f64 for cached headers, we now store the raw bytes of the uint256 work value.

Fixes #9

### CODE REVIEW

**Comment by mappum** [comment]
Nice, thanks for fixing this so quickly.

We'll have to note that this is a hard fork as it changes the serialization format and thus the Merkle roots. It was probably time to do a Tendermint rotation anyway since we're almost to 1M blocks on the current testnet.

On that note, we should also probably store versions with the data in the future so we can ensure we don't try to interpret old formats with a new struct. We should also create tests that will catch serialization changes (and other hard forks) so those changes don't take us by surprise. This could be as simple as test fixtures with known root hashes, or syncing the testnet and watching for apphash mismatches.

File: chain/chain/src/spv/headercache.rs [hunk 1]
```diff
@@ -173,7 +179,7 @@ impl<'a> HeaderCache<'a> {
     /// past.
     pub fn add_header_raw(&mut self, header: BlockHeader, height: u32) {
         let stored = StoredHeader {
-            log2work: Self::log2(header.work()),
+            work_bytes: header.work().to_bytes(),
```
Comment (mappum) [inline chain/chain/src/spv/headercache.rs]: Do we know `to_bytes` normalizes the byte order across architectures?

Comment (mappum) [reply chain/chain/src/spv/headercache.rs]: Ah, it returns an array of u64's, kind of a misleading name.

:+1:

**Comment by keppel** [comment]
> We'll have to note that this is a hard fork as it changes the serialization format and thus the Merkle roots.

True, so it's probably time to make a `develop` branch and stage breaking changes there.

**Comment by mappum** [comment]
The relayer test fails for me on this branch:
```
---- relayer::tests::run_relayer_state_machine stdout ----
did 100 relayer steps
thread 'relayer::tests::run_relayer_state_machine' panicked at 'called `Result::unwrap()` on an `Err` value: Error("missing field `work_bytes`", line: 1, column: 307)', chain/chain/src/spv/headercache.rs:257:25
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:77
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1052
   5: std::io::Write::write_fmt
             at /rustc/442ae7f04026c215a03b155eaaf9cde8bb5cf02a/src/libstd/io/mod.rs:1428
   6: std::io::impls::<impl std::io::Write for alloc::boxed::Box<W>>::write_fmt
             at src/libstd/io/impls.rs:156
   7: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   8: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   9: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:204
  10: std::panicking::default_hook
             at src/libstd/panicking.rs:221
  11: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:470
  12: rust_begin_unwind
             at src/libstd/panicking.rs:378
  13: core::panicking::panic_fmt
             at src/libcore/panicking.rs:85
  14: core::option::expect_none_failed
             at src/libcore/option.rs:1199
  15: core::result::Result<T,E>::unwrap
             at /rustc/442ae7f04026c215a03b155eaaf9cde8bb5cf02a/src/libcore/result.rs:963
  16: nomic_chain::spv::headercache::HeaderCache::get_header
             at chain/chain/src/spv/headercache.rs:257
  17: nomic_chain::spv::headercache::HeaderCache::tip
             at chain/chain/src/spv/headercache.rs:466
  18: nomic_client::client::Client::get_bitcoin_tip
             at chain/client/src/client.rs:109
  19: relayer::relayer::tests::run_relayer_state_machine
             at relayer/src/relayer.rs:329
  20: relayer::relayer::tests::run_relayer_state_machine::{{closure}}
             at relayer/src/relayer.rs:321
  21: core::ops::function::FnOnce::call_once
             at /rustc/442ae7f04026c215a03b155eaaf9cde8bb5cf02a/src/libcore/ops/function.rs:232
  22: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/442ae7f04026c215a03b155eaaf9cde8bb5cf02a/src/liballoc/boxed.rs:1016
  23: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:86
  24: std::panicking::try
             at /rustc/442ae7f04026c215a03b155eaaf9cde8bb5cf02a/src/libstd/panicking.rs:281
  25: std::panic::catch_unwind
             at /rustc/442ae7f04026c215a03b155eaaf9cde8bb5cf02a/src/libstd/panic.rs:394
  26: test::run_test_in_process
             at src/libtest/lib.rs:539
  27: test::run_test::run_test_inner::{{closure}}
             at src/libtest/lib.rs:452
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.


failures:
    relayer::tests::run_relayer_state_machine

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
```

**Comment by keppel** [comment]
That test currently assumes you've got the ABCI server, Tendermint, and a regtest bitcoin full node running locally. It's about time to add some real unit tests there.

**Comment by mappum** [comment]
If you're sure this change doesn't break anything, we could just add `#[ignore]` to that test and merge it, then add better tests later.

**Pull request merged**